selfdrive/common/clutil.cc

```diff
@@ -0,0 +1,212 @@
+  printf("name: '%s'\n", str);
+
+  clGetDeviceInfo(device, CL_DEVICE_VERSION, sizeof(str), str, NULL);
+  printf("device version: '%s'\n", str);
+
+  size_t sz = 0;
+  clGetDeviceInfo(device, CL_DEVICE_MAX_WORK_GROUP_SIZE, sizeof(sz), &sz, NULL);
+  printf("max work group size: %zu\n", sz);
+
+  cl_device_type type = 0;
+  clGetDeviceInfo(device, CL_DEVICE_TYPE, sizeof(type), &type, NULL);
+  printf("type = 0x%04x = ", (unsigned int)type);
+  switch (type) {
+    case CL_DEVICE_TYPE_CPU:
+      printf("CL_DEVICE_TYPE_CPU\n");
+      break;
+    case CL_DEVICE_TYPE_GPU:
+      printf("CL_DEVICE_TYPE_GPU\n");
+      break;
+    case CL_DEVICE_TYPE_ACCELERATOR:
+      printf("CL_DEVICE_TYPE_ACCELERATOR\n");
+      break;
+    default:
+      printf("Other...\n");
+      break;
+  }
+}
+
+void cl_print_build_errors(cl_program program, cl_device_id device) {
+  cl_build_status status;
+  clGetProgramBuildInfo(program, device, CL_PROGRAM_BUILD_STATUS,
+                        sizeof(cl_build_status), &status, NULL);
+
+  size_t log_size;
+  clGetProgramBuildInfo(program, device, CL_PROGRAM_BUILD_LOG, 0, NULL, &log_size);
+
+  std::unique_ptr<char[]> log = std::make_unique<char[]>(log_size + 1);
+  clGetProgramBuildInfo(program, device, CL_PROGRAM_BUILD_LOG, log_size + 1, &log[0], NULL);
+  printf("build failed; status=%d, log:\n%s\n", status, &log[0]);
+}
+
+}  // namespace
+
+cl_device_id cl_get_device_id(cl_device_type device_type) {
+  cl_uint num_platforms = 0;
+  CL_CHECK(clGetPlatformIDs(0, NULL, &num_platforms));
+  std::unique_ptr<cl_platform_id[]> platform_ids = std::make_unique<cl_platform_id[]>(num_platforms);
+  CL_CHECK(clGetPlatformIDs(num_platforms, &platform_ids[0], NULL));
+
+  for (size_t i = 0; i < num_platforms; i++) {
+    std::string platform_name = get_platform_info(platform_ids[i], CL_PLATFORM_NAME);
+    printf("platform[%zu] CL_PLATFORM_NAME: %s\n", i, platform_name.c_str());
+
+    cl_uint num_devices;
+    int err = clGetDeviceIDs(platform_ids[i], device_type, 0, NULL, &num_devices);
+    if (err != 0 || !num_devices) {
+      continue;
+    }
+    // Get first device
+    cl_device_id device_id = NULL;
+    CL_CHECK(clGetDeviceIDs(platform_ids[i], device_type, 1, &device_id, NULL));
+    cl_print_info(platform_ids[i], device_id);
+    return device_id;
+  }
+  printf("No valid openCL platform found\n");
+  assert(0);
+  return nullptr;
+}
+
+cl_program cl_program_from_file(cl_context ctx, cl_device_id device_id, const char* path, const char* args) {
+  char* src = (char*)read_file(path, nullptr);
+  assert(src != nullptr);
+  cl_program prg = CL_CHECK_ERR(clCreateProgramWithSource(ctx, 1, (const char**)&src, NULL, &err));
+  free(src);
+  if (int err = clBuildProgram(prg, 1, &device_id, args, NULL, NULL); err != 0) {
+    cl_print_build_errors(prg, device_id);
+    assert(0);
+  }
+  return prg;
+}
+
+// Given a cl code and return a string represenation
+const char* cl_get_error_string(int err) {
```

Comment:
Maybe use a macro like in this example: https://gist.github.com/allanmac/9328bb2d6a99b86883195f8f78fd1b93

Then we don't risk putting the wrong exit code with the wrong string.